### ISSUE TITLE
How to config consutomer to prevent the same message in a NATS stream would be read multiple times by the source connector

### ISSUE THREAD

**Issue by jorgen-zhao** [Chrisss93/flink-connector-nats#5]
Hi, I am a developer who is interested in how Flink can connect to Jetstream. You have done a great job.
I had cloned and tried this project. There is a thing that confused me. In your README.
 `When the connector is configured with multiple explicit consumers rather than a default consumer, each of the consumers' subject-filters is validated to make sure they are non-overlapping (or else the same message in a NATS stream would be read multiple times by the source connector)` 
So, how to config that. A message should be read once. please give some tips. thanks a lot.

**Comment by Chrisss93** [comment]
Hi there. That note is mainly for more bespoke usage. This is probably best explained with an example.

Lets say there is a NATS stream called `Foo` with the subject-filter: `foo.>`. If you create a Flink source for that NATS stream where you don't specify the subject-filter manually, there's nothing to worry about. i.e.

```java
JetStreamSource.<String>builder()
    .setServerURL("nats://...")
    .setStream("Foo")
    .setDefaultConsumerConfiguration("my-flink-job")
```

The connector allows you to manually configure how many consumers that the Flink source should use when reading from the NATS stream and what their subject-filters should be. This can allow more advanced fine-grained control over the source parallelism if you want it. Consider:

```java
JetStreamSource.<String>builder()
    .setServerURL("nats://...")
    .setStream("foo")
    .addConsumerConfiguration(
        ConsumerConfiguration.builder()
                .filterSubject("foo.bar")
                .name("first-flink-foo-consumer")
                .durable("first-flink-foo-consumer")
    )
    .addConsumerConfiguration(
        ConsumerConfiguration.builder()
                .filterSubject("foo.other")
                .name("second-flink-foo-consumer")
                .durable("second-flink-foo-consumer")
    )
```

This behaviour should be OK because `foo.bar` and `foo.other` are disjoint so they can never capture the same message and they are both subsets of the stream's subject-filter `foo.>`. You might miss some records in the NATS stream since your subject-filters might not exhaust all the possibilities of the NATS stream's subject filter, so you probably should only do this if you know your NATS message subject-names have some sort of predictable semantic structure.

Now consider a different source configuration:

```java
JetStreamSource.<String>builder()
    .setServerURL("nats://...")
    .setStream("foo")
    .addConsumerConfiguration(
        ConsumerConfiguration.builder()
                .filterSubject("foo.bar.*")
                .name("first-flink-foo-consumer")
                .durable("first-flink-foo-consumer")
    )
    .addConsumerConfiguration(
        ConsumerConfiguration.builder()
                .filterSubject("foo.*.person")
                .name("second-flink-foo-consumer")
                .durable("second-flink-foo-consumer")
    )
```

This is invalid because `foo.bar.*` and `foo.*.person` overlap. Any NATS messages on the `Foo` stream with the subject  `foo.bar.person` will be read by both the "first-flink-foo-consumer" and the "second-flink-foo-consumer". The builder will prevent you from instantiating a source connector with these sorts of overlapping consumers in order to avoid message duplication.

**TLDR: The source will NOT read a message on the NATS stream more than once in normal usage. And it WILL stop you if you accidentally make a source configuration like the above where message duplication is possible.**

**Comment by jorgen-zhao** [comment]
Thanks for your explanation in detail. It very can be helpful. 
Now I am continuing to do some more interesting examples using this connector. :)
Thanks again.
